packages/api/src/routes/metaplex-upload.js

```diff
@@ -0,0 +1,158 @@
+import * as cluster from '../cluster.js'
+import { secrets, database } from '../constants.js'
+import { JSONResponse } from '../utils/json-response.js'
+import { toNFTResponse } from '../utils/db-transforms.js'
+import { parseCid } from '../utils/utils.js'
+import {
+  parseJWTHeader,
+  parseJWT,
+  Base64URL,
+  utf8ToUint8Array,
+} from '../utils/jwt.js'
+import { keyFromDID, verifyEd25519Signature } from '../utils/ed25519.js'
+import { DBClient, DBError } from '../utils/db-client.js'
+
+/**
+ * When >2.5MB, use local add, because waiting for blocks to be sent to
+ * other cluster nodes can take a long time. Replication to other nodes
+ * will be done async by bitswap instead.
+ */
+const LOCAL_ADD_THRESHOLD = 1024 * 1024 * 2.5
+
+// TODO: remove when https://github.com/nftstorage/nft.storage/pull/638 merged
+const db = new DBClient(database.url, secrets.database)
+
+/**
+ * Temporary CAR only upload endpoint, authenticated with Metaplex JWT.
+ * @type {import('../bindings').Handler}
+ */
+export async function metaplexUpload(event, ctx) {
+  const { headers } = event.request
+  const authHeader = headers.get('x-web3auth') || ''
+
+  const match = authHeader.match(/^Metaplex (.*)$/)
+  if (!match) {
+    throw new HTTPError('invalid authorization header: ' + authHeader, 401)
+  }
+
+  const token = match[1]
+  const valid = await verifyMetaplexJWT(token)
+  if (!valid) {
+    throw new HTTPError('invalid authorization header', 401)
+  }
+
+  // now that we know the sig is valid, use the JWT payload
+  // as metadata to attach to the upload entry in the db
+  /** @type Record<string, any> */
+  const { iss, req } = parseJWT(token)
+  if (!iss) {
+    throw new HTTPError(
+      'invalid token - required field "iss" not present in payload',
+      401
+    )
+  }
+  if (!req) {
+    throw new HTTPError(
+      'invalid token - required field "req" not present in payload',
+      401
+    )
+  }
+
+  // Get the mapped metaplex user
+  const { user, key } = await validate()
+
+  const blob = await event.request.blob()
+  if (blob.size === 0) {
+    throw new HTTPError('Empty payload', 400)
+  }
+
+  const local = blob.size > LOCAL_ADD_THRESHOLD
+  const { cid, bytes: dagSize } = await cluster.addCar(blob, { local })
+  const { sourceCid, contentCid } = parseCid(cid)
+
+  // confirm that the content cid matches the cid from the signed metadata
+  if (!req.put || typeof req.put.rootCID !== 'string') {
+    throw new HTTPError('invalid token - no root cid in payload', 401)
+  }
+  const signedCid = req.put.rootCID
+  if (signedCid !== sourceCid && signedCid !== contentCid) {
+    throw new HTTPError('invalid token - cid mismatch', 401)
+  }
+
+  const upload = await db.createUpload({
+    mime_type: blob.type,
+    type: 'Car',
+    content_cid: contentCid,
+    source_cid: sourceCid,
+    dag_size: dagSize,
+    user_id: user.id,
+    files: [],
+    key_id: key.id,
+    meta: { iss, req: JSON.stringify(req) },
```

Comment:
Why stringify here? We cannot query the contents from Postgres if it's a string.

```suggestion
    meta: { iss, req: { put: { rootCID: req.put.rootCID } } },
```

...also just need to ensure we don't put anything else in the DB that we don't want there. Do you want to include `req.put.tags`? If yes can we validate it in any way?

Reply:
I changed things just now to extract the `solanaCluster` from the tags and put it into metadata. That's the only bit we might need in the future, apart from the issuer & cid, I think. Now `meta` just has those three fields and ignores everything else.